### PR TITLE
Stabilization and Refactor

### LINKED ISSUE #196: Button is recieving two stated events when stateful controls are nested.

### PR DESCRIPTION
**Pull request by Iheartweb**
Fixes #196, #195, #192, #190, #187, #186, #185, #184, #178, #173, #162, #147, #139, #137

### CODE REVIEW

File: README.markdown [hunk 1]
```diff
@@ -27,9 +27,7 @@
 Lu is developed using [nodejs](http://nodejs.org/) and [compass](http://compass-style.org/). To start contributing:
 
 1. Fork Lu
-2. Install <a href="http://compass-style.org/" target="_blank">Compass</a>
-3. Install <a href="http://nodejs.org/" target="_blank">Node</a>
-4. In a terminal from the Lu directory:```$npm install```
-5. Then:```$compass watch```
```
Comment (Jimgskoop) [inline README.markdown]: We no longer need Compass???

File: examples/accordion.html [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  <style type="text/css">
+    body {
+      width: 800px;
+      margin: 50px auto;
+    }
+    .container {
+      height: auto;
+    }
+    .lu-state-collapsed .container {
+      height: 0;
```
Comment (Jimgskoop) [inline examples/accordion.html:28]: Nice.  I had "display: none" here, right?`

File: scripts/libraries/class.js [hunk 3]
```diff
@@ -1,216 +0,0 @@
-/**
```
Comment (Jimgskoop) [inline scripts/libraries/class.js:1]: bye bye class!

File: scripts/libraries/fiber.js [hunk 4]
```diff
@@ -0,0 +1,233 @@
+/**
```
Comment (Jimgskoop) [inline scripts/libraries/fiber.js:1]: hello fiber!

Some of the documentation seems missing or incomplete.

File: scripts/libraries/fiber.js [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    child.__base__ = parent;
+
+     // Make this class extendable
+    child.extend = Fiber.extend;
+
+    return child;
+  };
+
+  /**
+   * @purpose Return a proxy object for accessing base methods with a given context
```
Comment (Jimgskoop) [inline scripts/libraries/fiber.js:91]: is @purpose a JSDoc tag?

File: scripts/libraries/ie-localstorage-json-shim.js [hunk 6]
```diff
@@ -1,26 +0,0 @@
-/*
```
Comment (Jimgskoop) [inline scripts/libraries/ie-localstorage-json-shim.js:1]: Why are we removing this?

File: scripts/libraries/inject.js [hunk 7]
```diff
@@ -16,78 +16,87 @@ express or implied.   See the License for the specific language
...
+/** @constant identifies a path as relative */
 var RELATIVE_PATH_REGEX = /^(\.{1,2}\/).+/;
 
-// identifies a require() path as absolute fully-qualified URL
+/** @constant identifies a path as absolute fully-qualified URL */
 var ABSOLUTE_PATH_REGEX = /^([A-Za-z]+:)?\/\//;
 
-// run a test to determine if localstorage is available
+/** @constant run a test to determine if localstorage is available */
 var HAS_LOCAL_STORAGE = (function() {
```
Comment (Jimgskoop) [inline scripts/libraries/inject.js:100]: all of these separate var statements make me woozy.

File: scripts/libraries/inject.js [hunk 8]
```diff
@@ -875,13 +892,16 @@ var Analyzer;
...
       extractRequires: function(file) {
         var requires = [];
+        var requireMatches = null;
+        var defines = null;
         var uniques = {};
-        var match;
         var dirtyRuntimeRequires = [];
+        var dirtyStaticRequires = [];
         var staticRequires = [];
+        var inlineAMD = {};
```
Comment (Jimgskoop) [inline scripts/libraries/inject.js:904]: too many var statements

File: scripts/libraries/inject.js [hunk 9]
```diff
@@ -899,7 +919,7 @@ var Analyzer;
         }
         if (dirtyRuntimeRequires.length > 0) {
           try {
-            eval(dirtyRuntimeRequires.join(";"));
+            eval([dirtyRuntimeRequires.join(";"), "//@ sourceURL=inject-analyzer.js"].join("\n"));
```
Comment (Jimgskoop) [inline scripts/libraries/inject.js:922]: Should "//@ sourceURL=inject-analyzer.js" be hardcoded?

File: scripts/libraries/inject.js [hunk 10]
```diff
@@ -1248,14 +1287,15 @@ var Executor;
...
+        var toExec = code.replace(/([\w\W]+?)=([\w\W]*})[\w\W]*?$/, "$1 = ($2)();");
         var relativeE;
         toExec = [toExec, sourceString].join("\n");
         if (!context.Inject.INTERNAL.execute[options.functionId]) {
           // there is nothing to run, so there must have been an uncaught
-          // syntax error (firefox).
+          // syntax error (firefox). 
           try {
-            try { eval("+"); } catch (ee) { relativeE = ee; } eval(toExec);
+            try { eval("+\n//@ sourceURL=inject-executor-line.js"); } catch (ee) { relativeE = ee; }
```
Comment (Jimgskoop) [inline scripts/libraries/inject.js:1297]: should //@ sourceURL=inject-executor-line.js be hardcoded?

File: sandbox.html [hunk 11]
```diff
@@ -21,115 +21,15 @@
...
-      <li>17</li>
-      <li>18</li>
-      <li>19</li>
-      <li>20</li>
-      <li>21</li>
-    </ul>
-    <button data-lu="Button:Previous" type="button">&laquo;</button>
-    <button data-lu="Button:Next" type="button">&raquo;</button>
+  <div data-lu="Container">
+    <button data-lu="Button">State</button>
```
Comment (Jimgskoop) [inline sandbox.html]: Do we need to specify states here in the config?

File: sandbox.html [hunk 12]
```diff
@@ -141,13 +41,20 @@
   <script src="/scripts/libraries/jquery-1.7.1.min.js"></script>
 
   <script src="/scripts/lu-config.js" type="text/javascript" charset="utf-8"></script>
+  <script src="/scripts/lu.0.3.js" type="text/javascript" charset="utf-8"></script>
```
Comment (Jimgskoop) [inline sandbox.html]: lu.0.3.0.js ?

File: sandbox.html [hunk 13]
```diff
@@ -141,13 +41,20 @@
...
+  <script src="/scripts/mappers.js" type="text/javascript" charset="utf-8"></script>
```
Comment (Jimgskoop) [inline sandbox.html:42]: mappers.0.1.0.js?

File: sandbox.html [hunk 14]
```diff
@@ -141,13 +41,20 @@
...
 
   <script type="text/javascript" charset="utf-8">
-    Inject.setExpires(0);
+
+    Inject.setExpires( 0 );
     Inject.clearCache();
-    Inject.setModuleRoot("http://localhost:1337/");
-    window.lu_debug = 5
-    require.ensure( ['/scripts/lu'], function() {} );
```
Comment (Jimgskoop) [inline sandbox.html:150]: can do require.run("/scripts/lu") if you don't need the callback function

File: scripts/lu-controls/Abstract.js [hunk 15]
```diff
@@ -39,8 +41,10 @@ Abstract = Class.extend( function( Class ){
    */
   function addToEventStore( event, method ){
     var eventStore = this.eventStore;
-    _.each( _.trim(event).split( /\s+/g ), function( item ){
-      eventStore[item] = { method: method };
+    _.each( _.trim( event ).split( /\s+/g ), function( item ){
```
Comment (Jimgskoop) [inline scripts/lu-controls/Abstract.js]: Might be more efficient to move the regexp out of the loop.

File: scripts/lu-controls/Abstract.js [hunk 16]
```diff
@@ -52,7 +56,7 @@ Abstract = Class.extend( function( Class ){
    */
   function removeFromEventStore( event ){
     var eventStore = this.eventStore;
-    _.each( _.trim(event).split( /\s+/g ), function( item ){
+    _.each( _.trim( event ).split( /\s+/g ), function( item ){
```
Comment (Jimgskoop) [inline scripts/lu-controls/Abstract.js]: Might be more efficient to move the regexp out of the loop.

File: scripts/lu-controls/Abstract.js [hunk 17]
```diff
@@ -68,25 +72,25 @@ Abstract = Class.extend( function( Class ){
      * @param {Object} settings Custom settings for this component
      */
     init: function( $element, settings ){
-      //Set up observers and notifiers
       var $observe,
         $notify;
 
       _.defaults( settings, defaults );
 
+      this.$element = $element;
```
Comment (Jimgskoop) [inline scripts/lu-controls/Abstract.js:80]: Are we using the "this" keyword again?

File: scripts/lu-controls/Button.js [hunk 18]
```diff
@@ -51,40 +53,69 @@ Button = Switch.extend( function( base ){
...
+            requirements.push( decorators.def );
+            break;
+          case 'select':
+            requirements.push( decorators.select );
+            break;
+          case 'state':
+            requirements.push( decorators.state );
+            break;
+          default:
+            throw new Error( 'Button decorator "' + action + '" does not exist!' );
```
Comment (Jimgskoop) [inline scripts/lu-controls/Button.js]: What should catch this error?  Should we also trigger a Lu "error" event?

File: scripts/lu-controls/Button/decorators/state.js [hunk 19]
```diff
@@ -0,0 +1,121 @@
+var constants = require( 'lu/constants' );
+
+function stateDecorator( settings ){
```
Comment (Jimgskoop) [inline scripts/lu-controls/Button/decorators/state.js]: missing doc block

File: scripts/lu-controls/Abstract.js (hunk 15 above)
Comment (Iheartweb) [inline scripts/lu-controls/Abstract.js]: Done.

File: scripts/lu-controls/Abstract.js (hunk 17 above)
Comment (Iheartweb) [inline scripts/lu-controls/Abstract.js:80]: this and self

Comment (Iheartweb) [inline scripts/lu-controls/Abstract.js:80]: that's the idomatic way!

File: scripts/lu-controls/Button.js (hunk 18 above)
Comment (Iheartweb) [inline scripts/lu-controls/Button.js]: Nah this is just for devs who try to do some wanky stuff.

File: .gitignore [hunk 20]
```diff
@@ -8,4 +8,3 @@ bin/
 sandbox.html
 build/compass_cache
 !build/bin/
-examples/
```
Comment (Iheartweb) [inline .gitignore:11]: Add examples folder.

File: examples/accordion.html (hunk 2 above)
Comment (Iheartweb) [inline examples/accordion.html:28]: height is for potential animation.

File: scripts/components/Button.js [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+  helpers = require( 'lu/helpers' ),
+  Switch = require( 'lu/Switch' ),
+  Fiber = require( 'Fiber' ),
+  Button;
+
+Button = Switch.extend( function( base ){
+  var defaults = {
+      on: 'click'
+    },
+    root = 'lu/Button/decorators/',
```
Comment (Iheartweb) [inline scripts/components/Button.js:18]: put this in constants

File: scripts/components/Carousel.js [hunk 22]
```diff
@@ -103,30 +106,31 @@ Carousel =  List.extend( function ( base ) {
...
       } );
 
-      this.on( [PAUSE_EVENT, NEXT_EVENT, PREVIOUS_EVENT, FIRST_EVENT, LAST_EVENT, SELECT_EVENT].join( ' ' ), function( event, item ){
+      this.on( [constants.events.PLAY, constants.events.NEXT, constants.events.PREVIOUS, constants.events.FIRST, constants.events.LAST, constants.events.SELECT].join( ' ' ), function( event, item ){
         event.stopPropagation();
         self.pause();
       } );
 
+
       // this.on( OUT_OF_BOUNDS_EVENT + '.' + NEXT_EVENT, function( event ){
```
Comment (Iheartweb) [inline scripts/components/Carousel.js:134]: should this be commented out?

**Pull request merged**